### PR TITLE
PERF: improve efficiency of `BaseMaskedArray.__setitem__`

### PR DESCRIPTION
**Pull request by alexreg**
This somewhat deals with #44172, though that won't be fully resolved until 2D `ExtensionArray`s are supported (per the comments there).

CC @jbrockmendel

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
we also need an asv for this

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -364,6 +365,17 @@ def map_string(s):
 
     _HANDLED_TYPES = (np.ndarray, numbers.Number, bool, np.bool_)
 
+    def __setitem__(self, key, value):
+        if lib.is_bool(value):
+            key = check_array_indexer(self, key)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/boolean.py]: Take a look at the test_setitem_foo tests in tests.arrays.test_datetimelike.  Something similar in either tests.arrays.masked_shared or tests.arrays.(integer|boolean|floating).test_indexing

**Comment by jreback** [comment]
@alexreg status here? pls merge master as well

**Comment by alexreg** [comment]
@jreback Sorry, haven't had much time lately. It would probably take me a while to properly wrap my head around the testing framework and write good tests. Ideally someone else could step in with that, but if not, maybe I'll get to it in a month or two.

I rebased on current master, however, and pushed that. (Nice and easy.0

**Comment by jbrockmendel** [comment]
> Ideally someone else could step in with that

None of us are going to write tests for you, but we will [help](https://github.com/pandas-dev/pandas/pull/44186#discussion_r740343732) you through getting the hang of writing them.

```
def test_descriptive_name():
    obj = pd.whatever()
    with pytest.raises(something, match=expected_exception_message):
         obj[0] = some_invalid_value_that_doesnt_raise_now_but_does_with_your_bugfix
```

**Pull request closed without merging**